File: android/app/src/main/java/za/co/dielys/ui/tasks/Reorder.kt
package za.co.dielys.ui.tasks

import androidx.compose.foundation.lazy.LazyListItemInfo
import androidx.compose.foundation.lazy.LazyListState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * Takes the item at [from] and puts it at [to], everything else closing up behind
 * it. Pure, and separate from the gesture, because this is the part that has to be
 * right: the drop reads the two neighbours out of the result and hands them to
 * `Position.between`, so an off-by-one here is a task that lands in the wrong place
 * on both phones.
 */
fun <T> List<T>.moved(
    from: Int,
    to: Int,
): List<T> {
    if (from == to || from !in indices || to !in indices) return this
    val copy = toMutableList()
    copy.add(to, copy.removeAt(from))
    return copy
}

/**
 * Whether the on-screen order a drag produced should still be shown instead of
 * what Room says.
 *
 * A drag reorders locally before the write completes, so for a moment there are
 * two answers. The draft wins until the database agrees with it — and is dropped
 * the moment the two are about different tasks, because then something else
 * happened (the other phone added a row, a sync landed) and the draft is stale
 * rather than merely ahead.
 */
fun draftStillWanted(
    draft: List<String>,
    stored: List<String>,
): Boolean = draft != stored && draft.toSet() == stored.toSet()

/**
 * Long-press drag for a `LazyColumn`, in the only two hundred lines it takes.
 *
 * There is no auto-scroll at the edges yet: a shopping list that runs past one
 * screen is worth handling, but not before the thing works at all.
 */
class ReorderState(
    private val listState: LazyListState,
) {
    /** Index of the row under the finger, or null when nothing is being dragged. */
    var draggingIndex by mutableStateOf<Int?>(null)
        private set

    private var accumulated by mutableFloatStateOf(0f)
    private var initialOffset by mutableIntStateOf(0)

    /**
     * Rows below this index are not reorderable — the completed section and its
     * heading share the `LazyColumn`, and their indices must never be a drop
     * target.
     */
    private var limit by mutableIntStateOf(0)

    /** Pixels to shift the dragged row by so it stays under the finger. */
    val draggingOffset: Float
        get() =
            draggingItem?.let { item -> (initialOffset + accumulated) - item.offset } ?: 0f

    private val draggingItem: LazyListItemInfo?
        get() = draggingIndex?.let { index -> itemAt(index) }

    /**
     * @param y where the long press landed, in pixels from the top of the list.
     * @param count how many rows from the top are reorderable.
     * @return the index picked up, or null if the press was not on a draggable row.
     */
    fun start(
        y: Float,
        count: Int,
    ): Int? {
        val picked =
            listState.layoutInfo.visibleItemsInfo.firstOrNull { item ->
                item.index < count && y >= item.offset && y <= item.offset + item.size
            } ?: return null

        draggingIndex = picked.index
        initialOffset = picked.offset
        accumulated = 0f
        limit = count
        return picked.index
    }

    /**
     * Moves by [dy] and, when the dragged row's middle has crossed into a
     * neighbour, reports the swap through [onMove]. Middle rather than edge: the
     * row only takes a place once it is more than half into it, which is what
     * stops a slow drag from oscillating between two positions.
     */
    fun drag(
        dy: Float,
        onMove: (Int, Int) -> Unit,
    ) {
        accumulated += dy
        val current = draggingItem ?: return
        val top = current.offset + draggingOffset
        val middle = top + current.size / 2f

        val target =
            listState.layoutInfo.visibleItemsInfo.firstOrNull { item ->
                item.index != current.index &&
                    item.index < limit &&
                    middle >= item.offset &&
                    middle <= item.offset + item.size
            } ?: return

        onMove(current.index, target.index)
        draggingIndex = target.index
    }

    fun stop() {
        draggingIndex = null
        accumulated = 0f
        initialOffset = 0
    }

    private fun itemAt(index: Int): LazyListItemInfo? =
        listState.layoutInfo.visibleItemsInfo.firstOrNull { it.index == index }
}

File: web/src/ui/reorder.ts
/**
 * The pure half of drag-to-reorder (`useDragReorder.ts` is the gesture) — a
 * port of `android/.../ui/reorder/Reorder.kt`'s `moved` and
 * `draftStillWanted`, plus the one decision the web gesture makes that
 * Compose's `LazyColumn` made for Android: which slot the dragged row is over.
 */

/** Takes the item at [from] and puts it at [to], everything else closing up behind it. */
export function moved<T>(list: readonly T[], from: number, to: number): T[] {
  if (from === to || from < 0 || to < 0 || from >= list.length || to >= list.length) {
    return [...list];
  }
  const copy = [...list];
  const [item] = copy.splice(from, 1);
  copy.splice(to, 0, item as T);
  return copy;
}

/**
 * The `(afterId, beforeId)` a row at [index] of [order] sits between — what
 * `moveList`/`useTaskBoard().move` expect, and `domain/position.ts` places
 * between. Null on either side is an end of the list.
 */
export function neighborsOf(
  order: readonly string[],
  index: number,
): { afterId: string | null; beforeId: string | null } {
  return {
    afterId: order[index - 1] ?? null,
    beforeId: order[index + 1] ?? null,
  };
}

/**
 * Which slot the dragged row belongs in, given where every row's middle sat
 * when the drag began and where the dragged row's middle is now. A row gives
 * up its place once the dragged middle passes its own middle — the web
 * equivalent of Android's "more than half into it", which is what stops a
 * slow drag oscillating between two slots.
 */
export function dropTarget(centers: readonly number[], from: number, middle: number): number {
  let target = 0;
  centers.forEach((center, index) => {
    if (index !== from && center < middle) target++;
  });
  return target;
}

/**
 * Whether the order a drag produced should still be shown instead of what the
 * server says. The draft wins until the server agrees with it, and is dropped
 * the moment the two are about different rows — something else happened (a
 * row added, a list shared away) and the draft is stale rather than ahead.
 */
export function draftStillWanted(draft: readonly string[], stored: readonly string[]): boolean {
  if (draft.length !== stored.length) return false;
  const storedSet = new Set(stored);
  return draft.some((id, i) => id !== stored[i]) && draft.every((id) => storedSet.has(id));
}
